### PR TITLE
Soilwat save output

### PR DESCRIPTION
**Pull request by dschlaep**

### CODE REVIEW

**Comment by dschlaep** [comment]
Variable `colnames_OUT` got deleted from `SW_Output.c`, it is required for function `SW_OUT_set_colnames` and for `rSOILWAT2`.

The function `SW_OUT_set_colnames` itself got deleted. This function is required by `rSOILWAT2` and would provide a unified `SOILWAT2`-central way of speaking about the same outputs with the same names.

**Comment by dschlaep** [comment]
`key2str` in `SW_Output.c` got converted back to `static`. This is not ok, because `rSOILWAT2` externalizes it.

**Comment by dschlaep** [comment]
Iterator `styp` local to function `str2stype` in file `SW_Output.c` got converted back from `IntUS` to `OutSum`. This is not ok, because not all compilers allow an `enum` type to be incremented by `++`.

**Comment by dschlaep** [comment]
`ncol_OUT[eSW_Runoff]` in function `SW_OUT_set_ncol` in file `SW_Output.c` has the correct 4 on master and not 3.

**Comment by dschlaep** [comment]
This branch re-introduced outdated `SW_Output[k].period`. This has been completed replaced by `timeSteps`.

**Comment by dschlaep** [comment]
Branch code re-introduced unnecessary if-tests on output periods `if (timeSteps[k][i] < 4)`, e.g., in `SW_OUT_close_files`. Any not allowed values in `timeSteps` should already be captured by the `switch` code. However, the branch re-verted code back to using `switch(i)` instead of the updated and correct `switch (timeSteps[k][i])`.

**Comment by dschlaep** [comment]
I don't understand why SWA is treated differently than anything else in `SOILWAT2`. Why are all the calculations in `SW_Output.c`? -- all the rest of `SOILWAT2 does its calculations in dedicated simulation code/modules; and `SW_Output.c` does only outputting (and a little bit converting where absolutely needed)

**Comment by BrendenBe1** [comment]
@dschlaep @kpalmqui 
I believe I have resolved all the issues for STEPWAT. I have merged master for both SOILWAT and STEPWAT into my branch, as recently as Tuesday so everything is up to date. I did have to fix some stuff in STEPWAT that needed to be changed with the veg arrays after the update to SOILWAT with the new veg array structure. I was hoping you could take a look at the newest changes and see if we can merge it into master. Also to check that the values are still looking correct after updating STEPWAT veg arrays. One thing I did not change was the location of the calculations for SWA. SWA calculations depend on swcbulk so I was trying to figure out where the swcbulk was done being calculated and then do the swa stuff there but I am having trouble following the calculations of swcbulk.

**Comment by dschlaep** [comment]
> SWA calculations depend on swcbulk so I was trying to figure out where the swcbulk was done being calculated and then do the swa stuff there but I am having trouble following the calculations of swcbulk

The simulation loops through simulation years with `SW_Control/SW_CTL_main()` which calls `SW_CTL_run_current_year()` for each year. The function `SW_CTL_run_current_year()` loops through each day of the year and calls `SW_SWC_water_flow()` which simulates the water balance for that day of year. At the end of each day, it calls `_collect_values()` which does the outputting/output-aggregation for that day. Thus, the current correct values for SWC(bulk) for a day are ready after the call to `SW_SWC_water_flow()`.

**Comment by dschlaep** [comment]
I get 48 warnings when compiling commit b370249f4ed1efc8d547b7ea6aaafe96172d81a6 -- up from 1 warning on master.

```
make bint_run
cc  -O3 -Wall -Wextra -pedantic -std=c11 -c SW_Main_lib.c SW_VegEstab.c SW_Control.c generic.c rands.c Times.c mymemory.c filefuncs.c SW_Files.c SW_Model.c SW_Site.c SW_SoilWater.c SW_Markov.c SW_Weather.c SW_Sky.c SW_Output.c SW_VegProd.c SW_Flow_lib.c SW_Flow.c SW_Carbon.c
SW_Output.c:844:6: warning: implicit declaration of function 'stat_Output_Daily_CSV_Summary' is invalid in C99
      [-Wimplicit-function-declaration]
                                        stat_Output_Daily_CSV_Summary(-1);
                                        ^
SW_Output.c:846:7: warning: implicit declaration of function 'stat_Output_Weekly_CSV_Summary' is invalid in C99
      [-Wimplicit-function-declaration]
                                        stat_Output_Weekly_CSV_Summary(-1);
                                        ^
SW_Output.c:848:6: warning: implicit declaration of function 'stat_Output_Monthly_CSV_Summary' is invalid in C99
      [-Wimplicit-function-declaration]
                                        stat_Output_Monthly_CSV_Summary(-1);
                                        ^
SW_Output.c:850:6: warning: implicit declaration of function 'stat_Output_Yearly_CSV_Summary' is invalid in C99
      [-Wimplicit-function-declaration]
                                        stat_Output_Yearly_CSV_Summary(-1);
                                        ^
SW_Output.c:786:7: warning: unused variable 'str' [-Wunused-variable]
        char str[MAX_FILENAMESIZE];
             ^
SW_Output.c:787:7: warning: unused variable 'prefix' [-Wunused-variable]
        char prefix[MAX_FILENAMESIZE];
             ^
SW_Output.c:1399:29: warning: incompatible pointer types passing 'char *[500]' to parameter of type 'char *' [-Wincompatible-pointer-types]
                                        populate_output_values(reg_file_vals_day, soil_file_vals_day, k, 1, 0);
                                                               ^~~~~~~~~~~~~~~~~
SW_Output.c:312:35: note: passing argument to parameter 'reg_file_array' here
void populate_output_values(char *reg_file_array, char *soil_file_array, int output_var, int year_out, int outstr_file);
                                  ^
SW_Output.c:1399:48: warning: incompatible pointer types passing 'char *[500]' to parameter of type 'char *' [-Wincompatible-pointer-types]
                                        populate_output_values(reg_file_vals_day, soil_file_vals_day, k, 1, 0);
                                                                                  ^~~~~~~~~~~~~~~~~~
SW_Output.c:312:57: note: passing argument to parameter 'soil_file_array' here
void populate_output_values(char *reg_file_array, char *soil_file_array, int output_var, int year_out, int outstr_file);
                                                        ^
SW_Output.c:1403:103: warning: format specifies type 'char *' but the argument has type 'char **' [-Wformat]
  ...fprintf(SW_Output_Files.fp_dy_avg, "%d%c%d%c%s\n", SW_Model.simyear, _Sep, SW_Model.doy, _Sep, reg_file_vals_day);
                                                 ~~                                                 ^~~~~~~~~~~~~~~~~
SW_Output.c:1407:108: warning: format specifies type 'char *' but the argument has type 'char **' [-Wformat]
  ...fprintf(SW_Output_Files.fp_dy_soil_avg, "%d%c%d%c%s\n", SW_Model.simyear, _Sep, SW_Model.doy, _Sep, soil_file_vals_day);
                                                      ~~                                                 ^~~~~~~~~~~~~~~~~~
SW_Output.c:1422:29: warning: incompatible pointer types passing 'char *[500]' to parameter of type 'char *' [-Wincompatible-pointer-types]
                                        populate_output_values(reg_file_vals_week, soil_file_vals_week, k, 2, 0);
                                                               ^~~~~~~~~~~~~~~~~~
SW_Output.c:312:35: note: passing argument to parameter 'reg_file_array' here
void populate_output_values(char *reg_file_array, char *soil_file_array, int output_var, int year_out, int outstr_file);
                                  ^
SW_Output.c:1422:49: warning: incompatible pointer types passing 'char *[500]' to parameter of type 'char *' [-Wincompatible-pointer-types]
                                        populate_output_values(reg_file_vals_week, soil_file_vals_week, k, 2, 0);
                                                                                   ^~~~~~~~~~~~~~~~~~~
SW_Output.c:312:57: note: passing argument to parameter 'soil_file_array' here
void populate_output_values(char *reg_file_array, char *soil_file_array, int output_var, int year_out, int outstr_file);
                                                        ^
SW_Output.c:1432:109: warning: format specifies type 'char *' but the argument has type 'char **' [-Wformat]
  ...fprintf(SW_Output_Files.fp_wk_soil_avg, "%d%c%d%c%s\n", SW_Model.simyear, _Sep, SW_Model.week, _Sep, soil_file_vals_week);
                                                      ~~                                                  ^~~~~~~~~~~~~~~~~~~
SW_Output.c:1436:104: warning: format specifies type 'char *' but the argument has type 'char **' [-Wformat]
  ...fprintf(SW_Output_Files.fp_wk_avg, "%d%c%d%c%s\n", SW_Model.simyear, _Sep, SW_Model.week, _Sep, reg_file_vals_week);
                                                 ~~                                                  ^~~~~~~~~~~~~~~~~~
SW_Output.c:1451:29: warning: incompatible pointer types passing 'char *[500]' to parameter of type 'char *' [-Wincompatible-pointer-types]
                                        populate_output_values(reg_file_vals_month, soil_file_vals_month, k, 3, 0);
                                                               ^~~~~~~~~~~~~~~~~~~
SW_Output.c:312:35: note: passing argument to parameter 'reg_file_array' here
void populate_output_values(char *reg_file_array, char *soil_file_array, int output_var, int year_out, int outstr_file);
                                  ^
SW_Output.c:1451:50: warning: incompatible pointer types passing 'char *[500]' to parameter of type 'char *' [-Wincompatible-pointer-types]
                                        populate_output_values(reg_file_vals_month, soil_file_vals_month, k, 3, 0);
                                                                                    ^~~~~~~~~~~~~~~~~~~~
SW_Output.c:312:57: note: passing argument to parameter 'soil_file_array' here
void populate_output_values(char *reg_file_array, char *soil_file_array, int output_var, int year_out, int outstr_file);
                                                        ^
SW_Output.c:1461:110: warning: format specifies type 'char *' but the argument has type 'char **' [-Wformat]
  ...fprintf(SW_Output_Files.fp_mo_soil_avg, "%d%c%d%c%s\n", SW_Model.simyear, _Sep, SW_Model.month, _Sep, soil_file_vals_month);
                                                      ~~                                                   ^~~~~~~~~~~~~~~~~~~~
SW_Output.c:1465:105: warning: format specifies type 'char *' but the argument has type 'char **' [-Wformat]
  ...fprintf(SW_Output_Files.fp_mo_avg, "%d%c%d%c%s\n", SW_Model.simyear, _Sep, SW_Model.month, _Sep, reg_file_vals_month);
                                                 ~~                                                   ^~~~~~~~~~~~~~~~~~~
SW_Output.c:1480:29: warning: incompatible pointer types passing 'char *[500]' to parameter of type 'char *' [-Wincompatible-pointer-types]
                                        populate_output_values(reg_file_vals_year, soil_file_vals_year, k, 4, 0);
                                                               ^~~~~~~~~~~~~~~~~~
SW_Output.c:312:35: note: passing argument to parameter 'reg_file_array' here
void populate_output_values(char *reg_file_array, char *soil_file_array, int output_var, int year_out, int outstr_file);
                                  ^
SW_Output.c:1480:49: warning: incompatible pointer types passing 'char *[500]' to parameter of type 'char *' [-Wincompatible-pointer-types]
                                        populate_output_values(reg_file_vals_year, soil_file_vals_year, k, 4, 0);
                                                                                   ^~~~~~~~~~~~~~~~~~~
SW_Output.c:312:57: note: passing argument to parameter 'soil_file_array' here
void populate_output_values(char *reg_file_array, char *soil_file_array, int output_var, int year_out, int outstr_file);
                                                        ^
SW_Output.c:1484:84: warning: format specifies type 'char *' but the argument has type 'char **' [-Wformat]
  ...fprintf(SW_Output_Files.fp_yr_soil_avg, "%d%c%s\n", SW_Model.simyear, _Sep, soil_file_vals_year);
                                                  ~~                             ^~~~~~~~~~~~~~~~~~~
SW_Output.c:1488:79: warning: format specifies type 'char *' but the argument has type 'char **' [-Wformat]
  ...fprintf(SW_Output_Files.fp_yr_avg, "%d%c%s\n", SW_Model.simyear, _Sep, reg_file_vals_year);
                                             ~~                             ^~~~~~~~~~~~~~~~~~
SW_Output.c:2970:15: warning: unused variable 'curr_crit_rank_index' [-Wunused-variable]
                                int j, k, curr_crit_rank_index;
                                          ^
SW_Output.c:3157:23: warning: unused variable 'sum_loop' [-Wunused-variable]
        int j,kv,id_prev_veg,sum_loop,curr_vegType,curr_crit_rank_index,kv_veg_type,prev_crit_veg_type,greater_veg_type;
                             ^
SW_Output.c:3157:11: warning: unused variable 'id_prev_veg' [-Wunused-variable]
        int j,kv,id_prev_veg,sum_loop,curr_vegType,curr_crit_rank_index,kv_veg_type,prev_crit_veg_type,greater_veg_type;
                 ^
SW_Output.c:3160:11: warning: unused variable 'l' [-Wunused-variable]
        LyrIndex l;
                 ^
SW_Output.c:3161:8: warning: unused variable 'val' [-Wunused-variable]
        RealD val = SW_MISSING;
              ^
SW_Output.c:3156:67: warning: unused parameter 'p' [-Wunused-parameter]
static void get_dSWAbulk(RealF swa_master[16][16][20], int i, int p, OutPeriod pd){
                                                                  ^
SW_Output.c:3156:80: warning: unused parameter 'pd' [-Wunused-parameter]
static void get_dSWAbulk(RealF swa_master[16][16][20], int i, int p, OutPeriod pd){
                                                                               ^
SW_Output.c:6620:6: warning: unused variable 'j' [-Wunused-variable]
        int j;
            ^
SW_Output.c:7281:12: warning: incompatible pointer types passing 'char *[5000]' to parameter of type 'char *' [-Wincompatible-pointer-types]
                                strcat(colHeadersSoil, storeCol); // concatenate variable to string
                                       ^~~~~~~~~~~~~~
/usr/include/secure/_string.h:131:27: note: expanded from macro 'strcat'
                __builtin___strcat_chk (dest, __VA_ARGS__, __darwin_obsz (dest))
                                        ^~~~
SW_Output.c:7416:12: warning: incompatible pointer types passing 'char *[5000]' to parameter of type 'char *' [-Wincompatible-pointer-types]
                                strcat(colHeaders, storeRegCol); // concatenate variable to string
                                       ^~~~~~~~~~
/usr/include/secure/_string.h:131:27: note: expanded from macro 'strcat'
                __builtin___strcat_chk (dest, __VA_ARGS__, __darwin_obsz (dest))
                                        ^~~~
SW_Output.c:7425:72: warning: format specifies type 'char *' but the argument has type 'char **' [-Wformat]
                                fprintf(soil_file, "%s%c%s%c%s\n", col1Head, _Sep, col2Head, _Sep, colHeadersSoil); // write columns to file
                                                            ~~                                     ^~~~~~~~~~~~~~
SW_Output.c:7426:75: warning: format specifies type 'char *' but the argument has type 'char **' [-Wformat]
                                fprintf(regular_file, "%s%c%s%c%s\n", col1Head, _Sep, col2Head, _Sep, colHeaders); // write columns to file
                                                               ~~                                     ^~~~~~~~~~
SW_Output.c:7431:72: warning: format specifies type 'char *' but the argument has type 'char **' [-Wformat]
                                fprintf(soil_file, "%s%c%s%c%s\n", col1Head, _Sep, col2Head, _Sep, colHeadersSoil); // write columns to file
                                                            ~~                                     ^~~~~~~~~~~~~~
SW_Output.c:7432:75: warning: format specifies type 'char *' but the argument has type 'char **' [-Wformat]
                                fprintf(regular_file, "%s%c%s%c%s\n", col1Head, _Sep, col2Head, _Sep, colHeaders); // write columns to file
                                                               ~~                                     ^~~~~~~~~~
SW_Output.c:7437:72: warning: format specifies type 'char *' but the argument has type 'char **' [-Wformat]
                                fprintf(soil_file, "%s%c%s%c%s\n", col1Head, _Sep, col2Head, _Sep, colHeadersSoil); // write columns to file
                                                            ~~                                     ^~~~~~~~~~~~~~
SW_Output.c:7438:75: warning: format specifies type 'char *' but the argument has type 'char **' [-Wformat]
                                fprintf(regular_file, "%s%c%s%c%s\n", col1Head, _Sep, col2Head, _Sep, colHeaders); // write columns to file
                                                               ~~                                     ^~~~~~~~~~
SW_Output.c:7442:52: warning: format specifies type 'char *' but the argument has type 'char **' [-Wformat]
                                fprintf(soil_file, "%s%c%s\n", col1Head, _Sep, colHeadersSoil); // write columns to file
                                                        ~~                     ^~~~~~~~~~~~~~
SW_Output.c:7443:55: warning: format specifies type 'char *' but the argument has type 'char **' [-Wformat]
                                fprintf(regular_file, "%s%c%s\n", col1Head, _Sep, colHeaders); // write columns to file
                                                           ~~                     ^~~~~~~~~~
SW_Output.c:7185:14: warning: unused variable 'cnames_eSW_PET' [-Wunused-variable]
        const char *cnames_eSW_PET[] = { "pet_cm" };
                    ^
SW_Output.c:7154:7: warning: unused variable 'ctemp' [-Wunused-variable]
        char ctemp[50];
             ^
SW_Output.c:7182:14: warning: unused variable 'cnames_eSW_SurfaceWater' [-Wunused-variable]
        const char *cnames_eSW_SurfaceWater[] = { "surfaceWater_cm" };
                    ^
SW_Output.c:7180:14: warning: unused variable 'cnames_eSW_SoilInf' [-Wunused-variable]
        const char *cnames_eSW_SoilInf[] = { "soil_inf" };
                    ^
SW_Output.c:7183:14: warning: unused variable 'cnames_add_eSW_EvapSurface' [-Wunused-variable]
        const char *cnames_add_eSW_EvapSurface[] = { "evap_surfaceWater" };
                    ^
SW_Output.c:7187:14: warning: unused variable 'cnames_eSW_DeepSWC' [-Wunused-variable]
        const char *cnames_eSW_DeepSWC[] = { "lowLayerDrain_cm" };
                    ^
SW_Output.c:7151:87: warning: unused parameter 'std_headers' [-Wunused-parameter]
void create_col_headers(int outFileTimestep, FILE *regular_file, FILE *soil_file, int std_headers){
                                                                                      ^
SW_Output.c:7264:16: warning: comparison of integers of different signs: 'int' and 'LyrIndex' (aka 'unsigned int') [-Wsign-compare]
                                        for(q=1; q<=SW_Site.n_layers; q++){
                                                 ~^ ~~~~~~~~~~~~~~~~
48 warnings generated.
```

**Comment by dschlaep** [comment]
I don't understand why output functions are added to the module `SW_Control`:
```
void stat_Output_Daily_CSV_Summary(int iteration);
void stat_Output_Weekly_CSV_Summary(int iteration);
void stat_Output_Monthly_CSV_Summary(int iteration);
void stat_Output_Yearly_CSV_Summary(int iteration);
```

These new functions have nothing to do with controlling the flow of the simulation model.

**Comment by dschlaep** [comment]
I don't understand why the code to `Get proper order for rank_SWPcrits` is inserted inside the function `SW_CTL_main()` in module `SW_Control`.

This code is dealing solely with values of `SW_VegProd` and belongs there.

**Comment by dschlaep** [comment]
Also the code to `Get proper order for rank_SWPcrits` uses hard-coded numbers, e.g.,
```
RealF tempArray[4], tempArrayUnsorted[4];
tempArray[0] = SW_VegProd.critSoilWater[0];
```


This should make use of the proper macros, e.g.,

```
RealF tempArray[NVEGTYPES], tempArrayUnsorted[NVEGTYPES];
ForEachVegType(k)
{
    tempArray[k] = SW_VegProd.critSoilWater[k];
    tempArrayUnsorted[k] = SW_VegProd.critSoilWater[k];
}
```
here and everywhere else, e.g., 
```
typedef struct {
...
RealD critSoilWater[4];
int rank_SWPcrits[5];
...
} SW_VEGPROD;
```

should become
```
typedef struct {
...
RealD critSoilWater[NVEGTYPES];
int rank_SWPcrits[NVEGTYPES + 1];
...
} SW_VEGPROD;
```
to make the intention obvious.

**Comment by dschlaep** [comment]
We have a macro for missing values; e.g., 
```
tempArrayUnsorted[innerLoop] = 100;
```
should become
```
tempArrayUnsorted[innerLoop] = SW_MISSING;
```

**Comment by dschlaep** [comment]
Why is line
```
SW_OUT_set_colnames(); // as of yet not used in SOILWAT2-standalone
```
deleted from `SW_Main/main()`?

**Comment by dschlaep** [comment]
Why are you restricting `SW_OUT_set_colnames` to rSOILWAT2?

This was an attempt that SOILWAT2 provides consistent output field names for all of its uses.
If this should go to rSOILWAT2 then argue why and move it that repository.

**Comment by dschlaep** [comment]
Explain in comments what the units are of `SW_VEGPROD.critSoilWater[]`? And why not use the established scheme of naming soil moisture-related variables?

```
RealD critSoilWater[4]; // storing values in same order as defined in rgroup.in (0=tree, 1=shrub, 2=grass, 3=forb)
```

**Comment by dschlaep** [comment]
Why is `SW_VEGPROD.rank_SWPcrits[]` and array of 5 and not `NVEGTYPES = 4`? I don't see in the code where the additional element is used. There is no explanation in the code comments.

**Comment by dschlaep** [comment]
To me, `SW_VEGPROD.rank_SWPcrits[]` and `SW_VEGPROD.critSoilWater[]` look extremely like vegetation type specific variables. Why are they not in the `typedef struct {} VegType`?

**Comment by dschlaep** [comment]
Why is 
```
+#ifdef RSOILWAT
+#include <R.h>
+#include <Rdefines.h>
+#include <Rconfig.h>
+#include <Rinternals.h>
+#endif
```
code re-inserted back for total no use to `rands.h`???

**Comment by dschlaep** [comment]
Why did you revert in` testing/Input/outsetup.in`
```
-# Output setup file for SOILWAT2
+# Output setup file for SOILWAT
```

**Comment by dschlaep** [comment]
You re-verted all of the updated comments in `testing/Input/outsetup.in`

It really looks like your merge conflict resolution is not appropriate.

**Comment by dschlaep** [comment]
Wouldn't it be much easier to write code if you defined the file pointers `fp_dy, fp_dy_soil, ...` in `SW_OUTPUT` as arrays of pointers?

**Comment by dschlaep** [comment]
Only STEPWAT2 uses the file pointers `fp_dy_avg,fp_dy_avg_soil, ...` in `SW_OUTPUT`? Why not wrap them in `ifdef STEPWAT` directives?

**Comment by dschlaep** [comment]
Why is

+#ifdef RSOILWAT
+#include <R.h>
+#include <Rdefines.h>
+#include <Rconfig.h>
+#include <Rinternals.h>
+#endif

code re-inserted back for total no use to `SW_Files.h`???

**Comment by dschlaep** [comment]
The last column of each csv-output file has a column name but no values.

**Comment by dschlaep** [comment]
Use consistent column names. Currently: `VWCBULK_1` but `SWA_Grass_Lyr_1`

--> use and update if necessary the function `SW_OUT_set_colnames`

**Comment by dschlaep** [comment]
I get a segmentation fault:
```
./testing/SOILWAT2 -d ./testing -f files.in
make: *** [bint_run] Segmentation fault: 11
```
I removed line `TIMESTEP dy yr` from `testing/Input/outsetup.in`.

I strongly suspect that this is because the function `SW_OUT_read` restricts the use of the local variable `useTimeStep` to `ifdef STEPWAT` even though on master this is handled for all uses of `SOILWAT2`. Another poor merge-conflict resolution?

**Comment by dschlaep** [comment]
Why are we still having tons of hard-coded numbers in `get_SWAbulk`? I thought we addressed this with the last round of comments?

e.g., 
```
		smallestCritVal = SW_VegProd.critSoilWater[SW_VegProd.rank_SWPcrits[0]];
		largestCritVal = SW_VegProd.critSoilWater[SW_VegProd.rank_SWPcrits[3]];
		RealF dSWA_bulk[16][16][20];
		RealF dSWA_bulk_repartioned[16][16][20];
		RealF dSWA_repartitioned_sum[16][16];

		// loop through each veg type to get dSWAbulk
		for(curr_vegType = 3; curr_vegType >= 0; curr_vegType--){ // go through each veg type and recalculate if necessary. starts at smallest
			curr_crit_rank_index = SW_VegProd.rank_SWPcrits[curr_vegType]; // get rank index for start of next loop
			veg_type_in_use = SW_VegProd.veg[curr_crit_rank_index].cov.fCover; // set veg type fraction here
...
```

Could become
```
		smallestCritVal = SW_VegProd.critSoilWater[SW_VegProd.rank_SWPcrits[0]];
		largestCritVal = SW_VegProd.critSoilWater[SW_VegProd.rank_SWPcrits[NVEGTYPES - 1]];
		int nveg2 = NVEGTYPES * NVEGTYPES;
                RealF dSWA_bulk[nveg2][nveg2][20];
		RealF dSWA_bulk_repartioned[nveg2][nveg2][20];
		RealF dSWA_repartitioned_sum[nveg2][nveg2];

		// loop through each veg type to get dSWAbulk
		ForEachVegTypeBottomUp(curr_vegType) { // go through each veg type and recalculate if necessary. starts at smallest
			curr_crit_rank_index = SW_VegProd.rank_SWPcrits[curr_vegType]; // get rank index for start of next loop
			veg_type_in_use = SW_VegProd.veg[curr_crit_rank_index].cov.fCover; // set veg type fraction here
...
```

**Comment by dschlaep** [comment]
Why are vegetation-type related variables in `get_SWAbulk()` being defined as arrays of 16 x 16 and 16 x 16 x 20 etc?
```
		RealF dSWA_bulk[16][16][20];
		RealF dSWA_bulk_repartioned[16][16][20];
		RealF dSWA_repartitioned_sum[16][16];
```

Their indices in the code appear to never get larger than `NVEGTYPES`?

**Comment by dschlaep** [comment]
This commit doesn't work with the master branch of `rSOILWAT2`:
```
1 error generated.
make: *** [SW_R_lib.o] Error 1
ERROR: compilation failed for package ‘rSOILWAT2’
```

**Comment by BrendenBe1** [comment]
I addressed most of the concerns in the recent commits.

#------------#

> I don't understand why output functions are added to the module SW_Control:

```
void stat_Output_Daily_CSV_Summary(int iteration);
void stat_Output_Weekly_CSV_Summary(int iteration);
void stat_Output_Monthly_CSV_Summary(int iteration);
void stat_Output_Yearly_CSV_Summary(int iteration);
```
>These new functions have nothing to do with controlling the flow of the simulation model.

I have moved these functions to filefuncs.c

#------------#

> I don't understand why the code to Get proper order for rank_SWPcrits is inserted inside the function SW_CTL_main() in module SW_Control.
> Also the code to Get proper order for rank_SWPcrits uses hard-coded numbers

I moved this calculation into SW_VegProd.c and replaced the hardcoded values.

#------------#

> Why are you restricting SW_OUT_set_colnames to rSOILWAT2?

I wrote a function called 'create_col_headers' that creates the column headers and I integrated some of the functionality of the new SW_OUT_set_colnames into it but I did not switch to using the new function because it does not do what needs to be done for the new output files. We can keep that functionality for use with rSOILWAT and move it there.

#------------#

> The last column of each csv-output file has a column name but no values.

Fixed this. There was a problem with both reg and soil files. In the regular file ESTAB was set in the outsetup.in file but it was off so it created a column header with no values. In the soil file there was an extra column created for evapsoil.

#------------#

> I get a segmentation fault:
> 
> ./testing/SOILWAT2 -d ./testing -f files.in
> make: *** [bint_run] Segmentation fault: 11
> I removed line TIMESTEP dy yr from testing/Input/outsetup.in.

This was because I changed the functionality of the output so it creates the files defined by TIMESTEP in the outsetup.in file so when you deleted the timestep line to use the period instead it gave a segfault because the files were not created. I solved this by adding a line in the outsetup.in file 'USE_TIMESTEP 1' that is the same as the outsetup file in STEPWAT. Setting this to 0 means to use the period and setting it to 1 is to use the timesteps.

#------------#

I also removed all the re-inserted code because it was not needed like you said.

File: SW_Control.c [hunk 1]
```diff
@@ -46,6 +46,7 @@ extern SW_MODEL SW_Model;
 extern SW_VEGESTAB SW_VegEstab;
 extern SW_SITE SW_Site;
 extern SW_VEGPROD SW_VegProd;
+extern SW_FILE_STATUS Sw_File_Status;
```
Comment (dschlaep) [inline SW_Control.c]: @BrendenBe1 Would you mind terribly to change the name of `Sw_File_Status` to match the naming scheme of every other variable in SOILWAT2? i.e., change it to `SW_File_Status`

Comment (BrendenBe1) [reply SW_Control.c]: Ya I will make that change

**Comment by BrendenBe1** [comment]
I removed the USE_TIMESTEP line from both SOILWAT and STEPWAT outsetup.in files. This functionality has been replaced in SW_Output.c with a check to see which timeperiods are defined and then creating the files for those timeperiods if not using TIMESTEP.

**Comment by dschlaep** [comment]
Thanks for handling `USE_TIMESTEP` simply as we used to.

**Comment by BrendenBe1** [comment]
@dschlaep @kpalmqui 
I moved the SWA calculations and was hoping you could take a look at them and see if the values are still accurate. I also got rSOILWAT2 to compile and install with this branch.

**Comment by dschlaep** [comment]
@BrendenBe1 Thanks for the update.
> I moved the SWA calculations and was hoping you could take a look at them and see if the values are still accurate.
--> Could you please produce SWA values for the first few days of `testing/` before and after?

> I also got rSOILWAT2 to compile and install with this branch.
This is good news. Did you run the unit tests of rSOILWAT2? In R: `devtools::tests()`

**Comment by BrendenBe1** [comment]
I attached two files with shrub values from SOILWAT2. The old value file is the output from SOILWAT2 just after merging master into my branch and the new values come from the most recent branch.

old commit (1370774b827d59dd71edbb8aceb48776f90741cf)



[old_shrub.txt](https://github.com/DrylandEcology/SOILWAT2/files/1742356/old_shrub.txt)
[new_Shrub.txt](https://github.com/DrylandEcology/SOILWAT2/files/1742358/new_Shrub.txt)

**Comment by kpalmqui** [comment]
The old_shrub.txt file contains values for forbs, not shrubs.

**Comment by BrendenBe1** [comment]
It is shrub values i just forgot to change the print statement to say shrubs instead of forbs.

**Comment by dschlaep** [comment]
Besides the labeling issue that you pointed out, why are the values different? Wasn't this supposed to be a code change and not a different algorithm? If so I expected identical output.

**Comment by BrendenBe1** [comment]
The values are different ever run.

**Comment by dschlaep** [comment]
This should not be because SOILWAT2 is deterministic. If they change from run to run, then this is a major warning sign that something fundamentally is wrong.

**Comment by kpalmqui** [comment]
I assume these were results produced within STEPWAT2? If so, the values shouldn't be the same every year, as the vegetation fractions/biomass/litter etc.changes year to year in STEPWAT2.

**Comment by dschlaep** [comment]
That may be. But this here is a pull-request in the SOILWAT2 repository and these are code changes to SOILWAT2. We need SOILWAT2 output to evaluate changes to the SOILWAT2 code.

**Comment by BrendenBe1** [comment]
@kpalmqui @dschlaep 
The latest set of issues has been dealt with and all unit tests for SOILWAT2, rSOILWAT2, and rSFSW2 are passing. I was hoping you could both review the latest commits.

**Pull request merged**